src/axom/core/Array.hpp

```diff
@@ -606,32 +614,62 @@ Array<T, DIM>::Array(Args... args)
 }
 
 //------------------------------------------------------------------------------
-template <typename T, int DIM>
-template <IndexType SFINAE, typename std::enable_if<SFINAE == 1>::type*>
-Array<T, DIM>::Array(IndexType num_elements, IndexType capacity, int allocator_id)
+template <typename T, int DIM, MemorySpace SPACE>
+template <IndexType SFINAE_DIM,
+          MemorySpace SFINAE_SPACE,
+          typename std::enable_if<SFINAE_DIM == 1 &&
+                                  SFINAE_SPACE == MemorySpace::Dynamic>::type*>
+Array<T, DIM, SPACE>::Array(IndexType num_elements,
+                            IndexType capacity,
+                            int allocator_id)
   : m_allocator_id(allocator_id)
 {
   initialize(num_elements, capacity);
 }
 
 //------------------------------------------------------------------------------
-template <typename T, int DIM>
-Array<T, DIM>::Array(const Array& other, int allocator_id)
-  : ArrayBase<T, DIM, Array<T, DIM>>(
-      static_cast<const ArrayBase<T, DIM, Array<T, DIM>>&>(other))
-  , m_allocator_id(allocator_id)
+template <typename T, int DIM, MemorySpace SPACE>
+template <IndexType SFINAE_DIM,
+          MemorySpace SFINAE_SPACE,
+          typename std::enable_if<SFINAE_DIM == 1 &&
+                                  SFINAE_SPACE != MemorySpace::Dynamic>::type*>
+Array<T, DIM, SPACE>::Array(IndexType num_elements, IndexType capacity)
+  : m_allocator_id(axom::detail::getAllocatorID<SPACE>())
 {
+  initialize(num_elements, capacity);
+}
+
+//------------------------------------------------------------------------------
+template <typename T, int DIM, MemorySpace SPACE>
+Array<T, DIM, SPACE>::Array(const Array& other, int allocator_id)
+  : ArrayBase<T, DIM, Array<T, DIM, SPACE>>(
+      static_cast<const ArrayBase<T, DIM, Array<T, DIM, SPACE>>&>(other))
+  , m_allocator_id(SPACE == MemorySpace::Dynamic
+                     ? allocator_id
```

Comment:
What if `allocator_id` is from the correct memory space, but a different pool?

Reply:
I'm not an Umpire expert by any means but I believe in this case we'd have to do some introspection on the allocator to try to track down its "parent allocator"?